File: ConceptOfTheDay/Python/src/anagram_substring_search.py
"""
@author Anirudh Sharma
"""


def isCountEqual(text_count, pattern_count):
    for i in range(0, 256):
        if text_count[i] != pattern_count[i]:
            return False
    return True


def find_indices(text, patten):
    # Lengths of input strings
    n = len(text)
    m = len(patten)
    # list that will store the indices
    indices = []
    # Frequency arrays - assuming we have a set of 256 characters
    text_count = [0] * 256
    pattern_count = [0] * 256
    # Loop for the first m characters
    for i in range(0, m):
        text_count[ord(text[i])] += 1
        pattern_count[ord(patten[i])] += 1
    # At this point, we have traversed m characters in both the arrays.
    # Now we will loop through the remaining characters
    for i in range(m, n):
        # Check if the counts of characters in frequency arrays are equal or not
        if isCountEqual(text_count, pattern_count):
            indices.append(i - m)
        # Discard the left most character
        text_count[ord(text[i - m])] -= 1
        # Include the current character
        text_count[ord(text[i])] += 1
    # For the last window
    if isCountEqual(text_count, pattern_count):
        indices.append(n - m)
    return indices


text = "BACDGABCDA"
pattern = "ABCD"
print(str(find_indices(text, pattern))[1:-1])

text = "XYYZXZYZXXYZ"
pattern = "XYZ"
print(str(find_indices(text, pattern))[1:-1])

File: ConceptOfTheDay/JavaScript/src/anagram_substring_search.js
/**
 * @author Anirudh Sharna
 */
function find_indices(s, p) {
    // Set of indices
    let result = [];
    // Build a hash of the letters.
    let wordHash = {}; // Master copy.
    let hash = {}; // Current copy.
    p.split('').forEach(function (letter) {
        wordHash[letter] = wordHash[letter] ? wordHash[letter] + 1 : 1;
        hash[letter] = hash[letter] ? hash[letter] + 1 : 1;
    });

    let count = 0;
    let index = -1;

    for (let i = 0; i < s.length; i++) {
        let letter = s[i];
        if (hash[letter]) {
            // Part or start of an anagram.
            if (index === -1) {
                index = i;
            }
            hash[letter]--;
            count++;
            if (count === p.length) {
                // Completed an anagram.
                result.push(index);

                // Reset variables.
                hash[s[index]]++;

                count = p.length - 1;
                index++;
            }
        }
        else {
            // Reset variables.
            if (index !== -1 && hash[letter] !== undefined && (p.length <= s.length - index)) {
                // Find first occurrence of 'letter'. Set the starting point 'index' to next letter.
                for (var j = index; j < i; j++) {
                    index++;
                    if (s[j] === letter) {
                        break;
                    }
                    else {
                        hash[s[j]]++;
                        count--;
                    }
                }
            }
            else {
                // Restore hash and counter.
                let keys = Object.keys(hash);
                for (let j = 0; j < keys.length; j++) {
                    hash[keys[j]] = wordHash[keys[j]];
                }
                count = 0;
                index = -1;
            }
        }
    }
    return result;
}


let text = "BACDGABCDA";
let pattern = "ABCD";
console.log("Anagrams found at: " + find_indices(text, pattern));

text = "XYYZXZYZXXYZ";
pattern = "XYZ";
console.log("Anagrams found at: " + find_indices(text, pattern));